crates/core/app/tests/mock_consensus_staking.rs

```diff
@@ -0,0 +1,196 @@
+    let existing_validator = match snapshot_end
+        .validator_definitions()
+        .tap(|_| info!("getting validator definitions"))
+        .await?
+        .as_slice()
+    {
+        [v] => v.clone(),
+        unexpected => panic!("there should be one validator, got: {unexpected:?}"),
+    };
+
+    let existing_validator_id = existing_validator.identity_key;
+
+    // Check that we are now in a new epoch.
+    {
+        use penumbra_sct::{component::clock::EpochRead, epoch::Epoch};
+        let Epoch { index: start, .. } = snapshot_start.get_current_epoch().await?;
+        let Epoch { index: end, .. } = snapshot_end.get_current_epoch().await?;
+        assert_eq!(start, 0, "we should start in the first epoch");
+        assert_eq!(end, 1, "we should now be in the second epoch");
+    }
+
+    // Show that the existing validator is and was active.
+    {
+        use penumbra_stake::{component::validator_handler::ValidatorDataRead, validator::State};
+        let start = snapshot_start
+            .get_validator_state(&existing_validator_id)
+            .await?;
+        let end = snapshot_end
+            .get_validator_state(&existing_validator_id)
+            .await?;
+        assert_eq!(start, Some(State::Active));
+        assert_eq!(end, Some(State::Active));
+    }
+
+    // Show that the validator was, and still is, in the consensus set.
+    {
+        use penumbra_stake::component::ConsensusIndexRead;
+        let start = snapshot_start.get_consensus_set().await?;
+        let end = snapshot_end.get_consensus_set().await?;
+        let expected = [existing_validator_id];
+        assert_eq!(
+            start, expected,
+            "validator should start in the consensus set"
+        );
+        assert_eq!(end, expected, "validator should stay in the consensus set");
+    }
+
+    // To define a validator, we need to define two keypairs: an identity key
+    // for the Penumbra application and a consensus key for cometbft.
+    let new_validator_id_sk = SigningKey::<SpendAuth>::new(OsRng);
+    let new_validator_id = IdentityKey(new_validator_id_sk.into());
+    let new_validator_consensus_sk = ed25519_consensus::SigningKey::new(OsRng);
+    let new_validator_consensus = new_validator_consensus_sk.verification_key();
+
+    // Insert the validator's consensus keypair into the keyring so it can be used to sign blocks.
+    /*
+    node.keyring_mut()
+        // Keyring should just be a BTreeMap rather than creating a new API
+        .insert(validator_consensus.clone(), validator_consensus_sk);
+     */
+
+    // Now define the validator's configuration data.
+    let new_validator = Validator {
+        identity_key: new_validator_id.clone(),
+        // TODO: upstream a direct conversion from ed25519_consensus
+        consensus_key: tendermint::PublicKey::from_raw_ed25519(&new_validator_consensus.to_bytes())
+            .expect("consensus key is valid"),
```

Comment:
filed https://github.com/informalsystems/tendermint-rs/pull/1401 to address this `TODO` :slightly_smiling_face: